plugins/backend-trace-forwarder/src/Cardano/BM/Backend/TraceForwarder.lhs

```diff
@@ -144,16 +144,30 @@ instance (ToJSON a) => IsEffectuator TraceForwarder a where
 
       writeMessageToQueue currentTF' = do
         let queue = tfQueue currentTF'
+        currentQueueSize <- atomically $ TBQ.lengthTBQueue queue
         noCapacity <- atomically $ TBQ.isFullTBQueue queue
-        if noCapacity
-          then do
-            let counterIORef = tfQueueFullCounter currentTF'
-            overflowed <- atomicModifyIORef' counterIORef $ \counter ->
-              if counter >= overflowCriticalNum
-                then (1, True)
-                else (counter + 1, False)
-            when overflowed $ handleOverflow tf
-          else atomically $ TBQ.writeTBQueue queue lo
+        if | noCapacity -> do
+             let counterIORef = tfQueueFullCounter currentTF'
+             overflowed <- atomicModifyIORef' counterIORef $ \counter ->
+               if counter >= overflowCriticalNum
+                 then (1, True)
+                 else (counter + 1, False)
+             when overflowed $ handleOverflow tf
+           | queueIsAlmostFull currentQueueSize -> do
+             -- Since the queue is almost full, it probably means that
+             -- the connection with acceptor is broken or too slow.
+             -- Remove current tfHandle to initiate establishing of
+             -- the new connection.
+             atomically $ modifyTVar' (getTF tf) $ \be -> be { tfHandle = Nothing }
+             -- Spawn new thread to establish new connection.
+             void $ Async.async $ establishConnection (getTF tf)
```

Comment:
We still should queue the message, though.

Reply:
Fixed.